Support more private keys types on PKCS8 format

BinAlyan opened the pull request:
Based on the previous pull request [here](https://github.com/SergioBenitez/Rocket/pull/1449). rustls already support different private key types, however they need to be in PKCS8 format [Check [here](https://github.com/ctz/rustls/pull/409#issuecomment-782300121)]. Adding support to them from Rocket is trivial.

I have added a tls_ecdsa example which use certifcate with signature scheme ECDSA_NISTP256_SHA256. The example includes a test, however it is copied from the TLS example so similarly it have this [issue](https://github.com/SergioBenitez/Rocket/issues/1566). The example ran successfully.

SergioBenitez commented:
Instead of creating a new example, can we simply set the keys as part of a new config profile in the existing `TLS` example? We should add an internal test for this in `http` so we don't depend on the example code.

BinAlyan commented:
In this pull request, I have added three new certificates with their keys and their gen scripts in the tls example. They can be run via Rocket profiles. They represent the currently supported signature schemes by rustls. They are as follows:
- ecdsa_nistp256_sha256
- ecdsa_nistp384_sha384
- ed2551

I have added two test functions which load these keys and certs. These two tests are unit tests for the functions load_private_key() and load_certs(). I am not sure if these two unit tests are enough or an integration test is required.

I load the keys and the certs from the tls example folder, I don't know if this is appropriate or kinda hacky. I am not sure of the macro I wrote because I never did Rust macros before.

SergioBenitez reviewed (commented):
Similarly with the generation scripts: why don't we have _one_ script that takes an argument controlling the kind of cert to generate?

BinAlyan commented:
> Similarly with the generation scripts: why don't we have _one_ script that takes an argument controlling the kind of cert to generate?

I am starting to make a single bash script generate all the certificates, however the bash script feel hacky and not that readable (to me), can I write the generation code using a rust code that invoke openssl via std::process::Command? I feel that would be simpler.

SergioBenitez commented:
Rust would be overkill for this. Just something like the following would be great:

```sh
function gen_ed25519() {
  openssl genpkey -algorithm ED25519 > ed25519_key.pem

  openssl req -new -key ed25519_key.pem -subj "${SUBJECT}" -out server.csr
  openssl x509 -req -extfile <(printf "subjectAltName=${ALT}") -days 3650 \
      -CA ca_cert.pem -CAkey ca_key.pem -CAcreateserial \
      -in server.csr -out ed25519_cert.pem

  rm ca_cert.srl server.csr
}

function gen_rsa_sha256() {
  openssl req -newkey rsa:4096 -nodes -sha256 \
    -keyout rsa_sha256_key.pem -subj "${SUBJECT}" -out server.csr

  openssl x509 -req -sha256 -extfile <(printf "subjectAltName=${ALT}") -days 3650 \
      -CA ca_cert.pem -CAkey ca_key.pem -CAcreateserial \
      -in server.csr -out rsa_sha256_cert.pem

  rm ca_cert.srl server.csr
}

case $1 in
  ed25519) gen_ed25519 ;;
  rsa_sha256) gen_rsa_sha256 ;;
  *)
    gen_ed25519
    gen_rsa_sha256
    ;;
esac
```

BinAlyan commented:
> Rust would be overkill for this. Just something like the following would be great:
> 
> ```shell
> function gen_ed25519() {
>   openssl genpkey -algorithm ED25519 > ed25519_key.pem
> 
>   openssl req -new -key ed25519_key.pem -subj "${SUBJECT}" -out server.csr
>   openssl x509 -req -extfile <(printf "subjectAltName=${ALT}") -days 3650 \
>       -CA ca_cert.pem -CAkey ca_key.pem -CAcreateserial \
>       -in server.csr -out ed25519_cert.pem
> 
>   rm ca_cert.srl server.csr
> }
> 
> function gen_rsa_sha256() {
>   openssl req -newkey rsa:4096 -nodes -sha256 \
>     -keyout rsa_sha256_key.pem -subj "${SUBJECT}" -out server.csr
> 
>   openssl x509 -req -sha256 -extfile <(printf "subjectAltName=${ALT}") -days 3650 \
>       -CA ca_cert.pem -CAkey ca_key.pem -CAcreateserial \
>       -in server.csr -out rsa_sha256_cert.pem
> 
>   rm ca_cert.srl server.csr
> }
> 
> case $1 in
>   ed25519) gen_ed25519 ;;
>   rsa_sha256) gen_rsa_sha256 ;;
>   *)
>     gen_ed25519
>     gen_rsa_sha256
>     ;;
> esac
> ```

I did not that was possible in bash. Thank you for your help. I have done the required changes.

SergioBenitez commented on core/http/src/tls.rs:
```diff
@@ -126,3 +126,45 @@ impl Connection for TlsStream<TcpStream> {
+        load_certs(&mut Cursor::new(ecdsa_nistp256_sha256_cert))?;
+        load_certs(&mut Cursor::new(ecdsa_nistp384_sha384_cert))?;
+        load_certs(&mut Cursor::new(ed2551_cert))?;
+
+
```
```suggestion

```

SergioBenitez commented on examples/tls/Rocket.toml:
```diff
@@ -1,9 +1,26 @@
+[rsa_sha256.tls]
+certs = "private/rsa_sha256_cert.pem"
+key = "private/rsa_sha256_key.pem"
+
+
```
```suggestion

```

SergioBenitez commented on examples/tls/private/gen_certs.sh:
```diff
@@ -0,0 +1,80 @@
+#! /bin/bash
+
+# Should use gen_ca.sh first to generate the CA.
```
Why don't we simply generate it here?

BinAlyan replied:
I separated the script for the case where the script user want to generate each certificate type separately. If the CA is generated each time, each CA generated will overwrite the previous CA. Should I add code that generates the CA only if it does not exist?

SergioBenitez replied:
I think so! We could add a "ca" case to the script that overwrites an existing CA cert (IE, does what the gen_ca.sh script does) and otherwise write out the CA cert if it doesn't exist.

SergioBenitez commented:
I've finished off the PR and will merge as soon as the CI passes.

SergioBenitez commented:
Merged in af48d1f2e64ac353511c079855951c9cdc51177a!

BinAlyan commented:
Thank you for completing the pull request and merging it!  Sorry for being late in completing this simple commit.

In core/http/src/tls.rs tests, I don't think the name of the macro **tls_example_key** is appropriate as it also was used for certs.

I think there is a typo in examples/tls/Rocket.toml:

> The certificate/private key pairs used here **where** generated via openssl using
